torch/csrc/jit/backends/backend.h

```diff
@@ -0,0 +1,397 @@
+          /*allow_any=*/true);
+
+      // This is a pointer to a backend instance that is used to access
+      // compile and execute functions.
+      auto cls = getCustomClass(qual_backend_name.qualifiedName());
+      TORCH_INTERNAL_ASSERT(cls);
+      c10::intrusive_ptr<torch::CustomClassHolder> backend;
+      loweredModule.register_attribute(
+          "__backend", cls, IValue::make_capsule(backend));
+
+      // This is the list of opaque backend handles returned by
+      // backend.compile.
+      loweredModule.register_attribute(
+          "__handles",
+          any_dict_ty,
+          c10::impl::GenericDict(
+              any_dict_ty->getKeyType(), any_dict_ty->getValueType()),
+          /*is_param=*/false,
+          /*allow_any=*/true);
+
+      // Methods.
+
+      // This is a helper function for creating a new instance of the
+      // backend class.
+      static const auto create_backend_ct = CodeTemplate(R"(
+            def __create_backend(self):
+                self.__backend = $name()
+            )");
+      TemplateEnv create_backend_te;
+      create_backend_te.s("name", qual_backend_name.qualifiedName());
+      loweredModule.define(
+          create_backend_ct.format(create_backend_te), resolver);
+
+      // getstate and setstate are for serialization/deserialization of the
+      // LoweredModule.
+      loweredModule.define(
+          R"(
+            def __getstate__(self):
+                return self.__method_compile_spec, self.__processed_module
+            )",
+          resolver);
+
+      loweredModule.define(
+          R"(
+            def __setstate__(self, state):
+                self.__method_compile_spec = state[0]
+                self.__processed_module = state[1]
+                self.__create_backend()
+                self.__handles = self.__backend.compile(self.__processed_module, self.__method_compile_spec)
+            )",
+          resolver);
+
+      // This is never called during compilation or execution, but is needed
+      // to generate the LoweredModule because we don't have access to an
+      // instance of the backend as a C++ object with which to call
+      // preprocess.
+      loweredModule.define(
+          R"(
+            def __preprocess(self, mod: Any, method_compile_spec: Dict[str, Any]):
+                self.__create_backend()
+                self.__processed_module = self.__backend.preprocess(mod, method_compile_spec)
+          )",
+          resolver);
+
+      // This loop generates one method on the LoweredModule for every key
+      // in method_compile_spec.
+      for (auto& e : method_compile_spec) {
+        std::string method_name = py::cast<std::string>(e.first);
+        static const auto method_ct = CodeTemplate(R"(
+            def $method(self${,def_inputs}):
+                typed_inputs: List[Any] = [${fwd_inputs,}]
+                $ret, = self.__backend.execute(self.__handles["$method"], typed_inputs)
+                return $ret
```

Comment:
you'll need to perform type refinement on the elements of `$ret`. Currently this is the generated code:
```
def forward(self,
    x: Tensor,
    h: Tensor) -> Tuple[Any, Any]:
  typed_inputs = annotate(List[Any], [x, h])
  _0 = (self.__backend).execute((self.__handles)["forward"], typed_inputs, )
  _1, _2, = _0
  return (_1, _2)
```
We have to add `isinstance(_1, Tensor)`, etc. to make sure that we are actually returning tensors.

Reply:
[followup] the isinstance approach won't work for any container types, e.g. `List[Tensor]`, because that's not valid python. We need to figure out how to support that. Also if we're checking container values, performance becomes a concern (as we ahve to do a linear inspection of all contained elements). An alternative is to make this an implicit requirement on backends and perform an unchecked cast (i.e. we're going to assume you handed us back the right type) but we should discuss.